src/client/partition.rs

```diff
@@ -295,24 +299,47 @@ impl BrokerCache for &PartitionClient {
             "Creating new partition-specific broker connection",
         );
 
-        let leader = self.get_leader(None).await?;
-        let broker = self.brokers.connect(leader).await?.ok_or_else(|| {
-            Error::InvalidResponse(format!(
-                "Partition leader {} not found in metadata response",
-                leader
-            ))
-        })?;
+        // Perform a request to fetch the topic metadata to determine the
+        // current leader.
+        //
+        // Accept a cached metadata entry as a response because:
+        //   * If the leader is offline, the connection call below fails and the cache is invalidated - the next call
+        //     will cause the cache to be re-populate with a fresh entry.
+        //   * A subsequent query is performed against the leader that does not use the cached entry - this validates
+        //     the correctness of the cached entry.
+        //
+        let leader = self.get_leader(None, true).await?;
+        let broker = match self.brokers.connect(leader).await {
+            Ok(Some(c)) => Ok(c),
+            Ok(None) => {
+                self.brokers.invalidate_metadata_cache();
+                Err(Error::InvalidResponse(format!(
+                    "Partition leader {} not found in metadata response",
+                    leader
+                )))
+            }
+            Err(e) => {
+                self.brokers.invalidate_metadata_cache();
+                Err(e.into())
+            }
+        }?;
 
         // Check if the chosen leader also thinks it is the leader.
         //
-        // For Kafka (+ Zookeeper) this seems to be required if we don't want to blindly retry
-        // `UnknownTopicOrPartition` that happen after we connect to a leader (as advised by another broker) which
-        // doesn't know about its assigned partition yet. The metadata query below seems to result in an
-        // LeaderNotAvailable in this case and is retried by the layer above.
+        // For Kafka (+ Zookeeper) this seems to be required if we don't want to blindly retry `UnknownTopicOrPartition`
+        // that happen after we connect to a leader (as advised by another broker) which doesn't know about its assigned
+        // partition yet. The metadata query below seems to result in an LeaderNotAvailable in this case and is retried
+        // by the layer above.
         //
         // This does not seem to be required for redpanda.
-        let leader_self = self.get_leader(Some(Arc::clone(&broker))).await?;
+        //
+        // Because this check requires the most up-to-date metadata from a specific broker, do not accept a cached
+        // metadata response.
+        let leader_self = self.get_leader(Some(Arc::clone(&broker)), false).await?;
```

Comment:
This check is the remaining request in the discovery phase and is somewhat unavoidable as-is.

I wonder if we can optimistically assume the leader is correct, trap on the `UnknownTopicOrPartition` error described here, and use it to trigger a re-run of discovery (with fresh metadata) to eliminate this pessimistic check?

Reply:
I think there is a similar question for #147. There we read from a broker that knows the partition (hence there is no `UnknownTopicOrPartition`) but is only a read replicate and hasn't caught up with the partition leader yet (i.e. we get an `OffsetOutOfRange`).

I agree that we should probably fix both by trapping these errors and assume that the user didn't want to run into them deliberately, invalidate the cache and retry. However we don't wanna retry forever because they are valid actual errors. So this needs a bit more sophisticated retry logic and I suggest we do that in a later PR.

Reply:
Yeah definitely not wanting to do it here...

I think ultimately performing an optimistic leader discovery would be much better for multiple reasons, but the subtleties sound tricky. It sounds like simply bounding the number of retries to something sensible might be an easy solution. I'll create a ticket and see how much pain is left in prod to see if it justifies working on it :+1: